feat: operators version service

taturosati opened the pull request:
Add operator version tracker

Exposes two endpoints:

- POST to /versions for operator to post its version on startup, verifies operator signature
- GET to /versions to list operator versions

# To test

Start anvil and aggregator

Start operator tracker: `make tracker_devnet_start`

Register operator: `make operator_full_registration`
Build operator: `make build_operator`
Start operator: `./operator/build/aligned-operator start --config ./config-files/config.yaml`

> [!WARNING]  
> Note that you can't use `make operator_start` because operator version is not included in executable if not built.

You can then get operator versions by running `curl http://localhost:3000/versions`

You can test an update by changing OPERATOR_VERSION in makefile, rebuilding and restarting operator

taturosati committed:
feat: operator post version + signature verification

taturosati committed:
feat: save version on db

taturosati committed:
refactor: rename operator_metrics to operator_tracker

taturosati committed:
feat: config from .env

taturosati committed:
feat: postgres dockerfile

taturosati committed:
fix: lint

JuArce commented on Makefile:
```diff
@@ -664,3 +664,33 @@ recover_db: run_db
+
+tracker_devnet_start: tracker_run_db
+	@cd operator_tracker/ && \
+		cp .env.dev .env && \
```
would be nice to run this target using the .env.dev file instead of overwriting the .env file

JuArce commented on config-files/config.yaml:
```diff
@@ -39,6 +39,7 @@ aggregator:
 ## Operator Configurations
 operator:
   aggregator_rpc_server_ip_port_address: localhost:8090
+  operator_tracker_ip_port_address: http://localhost:3000
```
This variable `operator_tracker_ip_port_address: http://localhost:3000` is missing in
- config-operator.yaml
- config-operator-1.yaml
- config-operator-2.yaml
- config-operator-3.yaml

JuArce commented on operator_tracker/sql/schema.sql:
```diff
@@ -0,0 +1,4 @@
+CREATE TABLE IF NOT EXISTS operator_versions (
+  address CHAR(42) PRIMARY KEY,
+  version TEXT
```
version should have a fixed length to avoid accept too long strings

JuArce commented on operator_tracker/src/main.rs:
```diff
@@ -0,0 +1,252 @@
+
+    let router = Router::new()
+        .route("/versions", post(post_operator_version))
+        .route("/versions", get(list_operator_versions))
```
There is missing a `/versions/:address`. Mainly for operator view in explorer to avoid request all versions every time

JuArce commented on operator_tracker/src/main.rs:
```diff
@@ -0,0 +1,252 @@
+        .await
+        .expect("Failed to execute query");
+
+    Json(rows)
```
Would be nice return 204 if there is no body content

JuArce commented on operator_tracker/src/main.rs:
```diff
@@ -0,0 +1,252 @@
+    }
+}
+
+async fn post_operator_version(
```
This should return 201 if it is created correctly. Also should include a link to the entity in the `Location` header and the created entity as response body

taturosati committed:
chore: run with .env.dev

taturosati committed:
feat: no content status code

taturosati committed:
feat (operator_tracker): no content / created status code + lib

taturosati committed:
feat (operator_tracker): get operator version

entropidelic commented on operator/cmd/actions/start.go:
```diff
@@ -34,6 +39,42 @@ func operatorMain(ctx *cli.Context) error {
+		return err
+	}
+
+	// send version to metrics server
```
small nit but this comment confused me 😛. It should be the operator tracker server

entropidelic commented on operator/cmd/actions/start.go:
```diff
@@ -34,6 +39,42 @@ func operatorMain(ctx *cli.Context) error {
+
+	// send version to metrics server
+	endpoint := operatorConfig.Operator.OperatorTrackerIpPortAddress + "/versions"
+	log.Println("Sending version to metrics server: ", endpoint)
```
same here. Also, can we use the eigensdk logger?

taturosati committed:
fix: check valid address

taturosati committed:
fix: operator check correct status codes

taturosati committed:
chore: update config files

entropidelic commented on operator_tracker/src/main.rs:
```diff
@@ -0,0 +1,154 @@
+use sqlx::postgres::{PgPool, PgPoolOptions};
+use sqlx::Executor;
+
+// The RegistryCoordinator contract
+// This is the contract that the service will interact with to get the status of the operator
+// The response is a uint8 that represents the status of the operator
+// 0 - Never registered
+// 1 - Registered
+// 2 - Deregistered
```
I don't really know how ethers handles this but couldn't we just use the contract enum?

taturosati replied:
unfortunately I don't think so, at least I could not find a way

taturosati committed:
chore: set max version length on db

entropidelic commented on operator_tracker/src/lib.rs:
```diff
@@ -0,0 +1,160 @@
+        axum::http::Response::builder()
+            .status(axum::http::StatusCode::BAD_REQUEST)
+            .body(Body::from(serde_json::to_string(&self).unwrap()))
+            .unwrap()
```
let's not use unwraps, if there is any error while building the response body let's respond something

entropidelic commented on operator_tracker/src/lib.rs:
```diff
@@ -0,0 +1,160 @@
+        axum::http::Response::builder()
+            .status(axum::http::StatusCode::OK)
+            .body(Body::from(serde_json::to_string(&self).unwrap()))
+            .unwrap()
```
same comment about responses here

taturosati committed:
refactor: use eigen logger

entropidelic commented on operator_tracker/src/lib.rs:
```diff
@@ -0,0 +1,175 @@
+    let status = registry_coordinator
+        .get_operator_status(operator_address)
+        .await
+        .expect("Failed to get operator status");
```
let's return an error here

entropidelic commented on operator_tracker/src/lib.rs:
```diff
@@ -0,0 +1,175 @@
+    // Recover operator address from signature
+
+    // hash keccak256(version) and recover address from signature
+    let version = keccak256(payload.version.as_bytes());
```
```suggestion
    let hashed_version = keccak256(payload.version.as_bytes());
```

entropidelic commented on operator_tracker/src/lib.rs:
```diff
@@ -0,0 +1,175 @@
+
+    db.execute(query.bind(&operator_address).bind(&payload.version))
+        .await
+        .expect("Failed to update operator version");
```
return error here

entropidelic commented on operator_tracker/src/lib.rs:
```diff
@@ -0,0 +1,175 @@
+        .bind(address)
+        .fetch_optional(db)
+        .await
+        .expect("Failed to execute query")
```
return error

entropidelic commented on operator_tracker/src/main.rs:
```diff
@@ -0,0 +1,170 @@
+            axum::http::Response::builder()
+                .status(axum::http::StatusCode::CREATED)
+                .body(Body::from(body))
+                .expect("Failed to build response")
```
same comment about responses when something fails. Let's fix that in all the response builders

taturosati committed:
fix: remove expects & unwraps

taturosati committed:
refactor: rename version to hashed_version

taturosati committed:
feat: location header

JuArce reviewed (commented):
Just one comment 
If the operator receives an error, it prints something like that

```
2024-08-21T12:14:58.066-0300	ERROR	actions/start.go:75	Error sending version to metrics server: 	{"err": null}
github.com/yetanotherco/aligned_layer/operator/cmd/actions.operatorMain
	/Users/juarce/repos/yetanothercompany/aligned_layer/operator/cmd/actions/start.go:75
github.com/urfave/cli/v2.(*Command).Run
	/Users/juarce/go/pkg/mod/github.com/urfave/cli/v2@v2.27.1/command.go:279
github.com/urfave/cli/v2.(*Command).Run
	/Users/juarce/go/pkg/mod/github.com/urfave/cli/v2@v2.27.1/command.go:272
github.com/urfave/cli/v2.(*App).RunContext
	/Users/juarce/go/pkg/mod/github.com/urfave/cli/v2@v2.27.1/app.go:337
github.com/urfave/cli/v2.(*App).Run
	/Users/juarce/go/pkg/mod/github.com/urfave/cli/v2@v2.27.1/app.go:311
main.main
	/Users/juarce/repos/yetanothercompany/aligned_layer/operator/cmd/main.go:27
runtime.main
	/opt/homebrew/Cellar/go/1.22.3/libexec/src/runtime/proc.go:271
```

Would be better to print a more useful message
It happens if the tracker service is not running or if the operator does not send a valid version (ie running without build)

taturosati committed:
refactor: change logs to warn

taturosati commented:
> Just one comment If the operator receives an error, it prints something like that
> 
> ```
> 2024-08-21T12:14:58.066-0300	ERROR	actions/start.go:75	Error sending version to metrics server: 	{"err": null}
> github.com/yetanotherco/aligned_layer/operator/cmd/actions.operatorMain
> 	/Users/juarce/repos/yetanothercompany/aligned_layer/operator/cmd/actions/start.go:75
> github.com/urfave/cli/v2.(*Command).Run
> 	/Users/juarce/go/pkg/mod/github.com/urfave/cli/v2@v2.27.1/command.go:279
> github.com/urfave/cli/v2.(*Command).Run
> 	/Users/juarce/go/pkg/mod/github.com/urfave/cli/v2@v2.27.1/command.go:272
> github.com/urfave/cli/v2.(*App).RunContext
> 	/Users/juarce/go/pkg/mod/github.com/urfave/cli/v2@v2.27.1/app.go:337
> github.com/urfave/cli/v2.(*App).Run
> 	/Users/juarce/go/pkg/mod/github.com/urfave/cli/v2@v2.27.1/app.go:311
> main.main
> 	/Users/juarce/repos/yetanothercompany/aligned_layer/operator/cmd/main.go:27
> runtime.main
> 	/opt/homebrew/Cellar/go/1.22.3/libexec/src/runtime/proc.go:271
> ```
> 
> Would be better to print a more useful message It happens if the tracker service is not running or if the operator does not send a valid version (ie running without build)

Fixed, changed log to warn and removed nil error

uri-99 commented on config-files/config.yaml:
```diff
@@ -39,6 +39,7 @@ aggregator:
 ## Operator Configurations
 operator:
   aggregator_rpc_server_ip_port_address: localhost:8090
+  operator_tracker_ip_port_address: http://localhost:3000
```
We are already using port 3000 for metrics,
found in 6_setup_aligned.md:
```
To run Prometheus and Grafana, run:

`make run_metrics`

Then you can access Grafana on `http://localhost:3000` with the default credentials `admin:admin`.
```

One solution could be to change the port to another one, 
Or a better solution could be to make the operator_version_tracker port, variable.

uri-99 replied:
port changed to 3030

uri-99 commented on operator_tracker/src/main.rs:
```diff
@@ -0,0 +1,178 @@
+
+    let rpc_url = &std::env::var("RPC_URL").expect("RPC_URL must be set");
+
+    let registry_coordinator_addr = &std::env::var("REGISTRY_COORDINATOR_ADDRESS")
+        .expect("REGISTRY_COORDINATOR_ADDRESS must be set");
```
We should consider reading this value from `alignedlayer_deployment_output.json`

taturosati replied:
I don't think its important, we would still need to pass path to the deployment output file, and this way we avoid the extra complexity of parsing the file

uri-99 commented on operator_tracker/src/main.rs:
```diff
@@ -0,0 +1,178 @@
+        .unwrap_or_default() // Should never fail but dont panic
+}
+
+async fn get_operator_version(
```
There should be a special response if there are no operator versions found, to avoid the following:

<img width="648" alt="image" src="https://github.com/user-attachments/assets/8e1fb7b2-739e-43ee-976d-96235292b692">

uri-99 replied:
will return 204 -no response- code

uri-99 commented on Makefile:
```diff
@@ -664,3 +664,32 @@ recover_db: run_db
+
+tracker_run_db: tracker_build_db tracker_remove_db_container
+	@cd operator_tracker && \
+		docker run -d --name tracker-postgres-container -p 5432:5432 -v tracker-postgres-data:/var/lib/postgresql/data tracker-postgres-image
```
5432 is also exposed for the Explorer DB, it will probably bring some issues on stage, but probably not on prod.

uri-99 replied:
port was changed to 5433

taturosati committed:
feat: add configurable listen address + change db port

taturosati committed:
fix: change tracker port on config files

entropidelic commented on operator/cmd/actions/start.go:
```diff
@@ -34,7 +39,45 @@ func operatorMain(ctx *cli.Context) error {
+	res, err := http.Post(endpoint, "application/json",
+		bodyBuffer)
+	if err != nil {
+		// Dont prevent operator from starting if metrics server is down
```
```suggestion
		// Dont prevent operator from starting if operator tracker server is down
```

entropidelic committed:
Update operator/cmd/actions/start.go

entropidelic commented on operator/cmd/actions/start.go:
```diff
@@ -34,7 +39,45 @@ func operatorMain(ctx *cli.Context) error {
+		// Dont prevent operator from starting if operator tracker server is down
+		operator.Logger.Warn("Error sending version to metrics server: ", "err", err)
+	} else if res.StatusCode != http.StatusCreated && res.StatusCode != http.StatusNoContent {
+		operator.Logger.Warn("Error sending version to metrics server: ", "status_code", res.StatusCode)
```
```suggestion
		operator.Logger.Warn("Error sending version to operator tracker server: ", "status_code", res.StatusCode)
```

entropidelic committed:
Update operator/cmd/actions/start.go
